upf/upf.c

```diff
@@ -201,44 +201,51 @@ vnet_upf_upip_add_del (ip4_address_t * ip4, ip6_address_t * ip6,
 {
   upf_main_t *gtm = &upf_main;
   upf_upip_res_t *ip_res;
-  upf_upip_res_t res = {
-    .ip4 = *ip4,
-    .ip6 = *ip6,
-    .nwi_index = ~0,
-    .intf = intf,
-    .teid = teid,
-    .mask = mask
-  };
+  u8 *key = 0;
   uword *p;
 
-  if (name)
+  if (!name)
     {
-      p = hash_get_mem (gtm->nwi_index_by_name, name);
-      if (!p)
-	return VNET_API_ERROR_NO_SUCH_ENTRY;
-
-      res.nwi_index = p[0];
+      upf_debug ("NWI is mandatory for GTPU endpoint");
+      return -1;
     }
 
-  p = mhash_get (&gtm->upip_res_index, &res);
+  /*
+   * TODO: For some reason pure usage of binary vector breaks
+   * hash lookup by key. Converting to string works like a
+   * workaround, but this has to be investigated and fixed.
+   */
+  name = format (0, "%s", name);
+
+  p = hash_get_mem (gtm->upip_res_index_by_nwi, name);
 
   if (add)
     {
       if (p)
 	return VNET_API_ERROR_VALUE_EXIST;
 
       pool_get (gtm->upip_res, ip_res);
-      memcpy (ip_res, &res, sizeof (res));
+      ip_res->teid = teid;
+      ip_res->intf = intf;
+      ip_res->ip4.as_u32 = ip4->as_u32;
+      ip_res->ip6.as_u64[0] = ip6->as_u64[0];
+      ip_res->ip6.as_u64[1] = ip6->as_u64[1];
+      ip_res->mask = mask;
+
+      p = hash_get_mem (gtm->nwi_index_by_name, name);
+      if (!p)
+	return VNET_API_ERROR_NO_SUCH_ENTRY;
+      ip_res->nwi_index = p[0];
 
-      mhash_set (&gtm->upip_res_index, ip_res, ip_res - gtm->upip_res, NULL);
+      hash_set_mem (gtm->upip_res_index_by_nwi, name, ip_res - gtm->upip_res);
```

Comment:
The use of name in the hash is broken and might even leak memory.

At this point in the code, `name` is a vector that has no owner. That mean nothing will be freeing it (the format call in line #218 creates a new vector). Name can not be freed in here, since the hash requires it to stay valid as long as the entry is in the hash. The core around hash_unset_mem would need to free it, but has not access to it.

Instead of attempting to build broken fixes to the UE IP resource pool thing, mak Rel.16 the default and just drop it completely!